### PR TITLE
chore: [dev-workspace] Rework SSH plugin to store keys in the K8s secret

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
* Rework SSH plugin to store keys in the K8s secret
* Generate SSH keys using the container tools instead of CHE API requests

Todo: uncomment the GitHub Communds when https://github.com/eclipse/che/issues/20860 is done.
### What issues does this PR fix or reference?

fixes https://github.com/eclipse/che/issues/20859

### How to test this PR?

1. Start the dev-workspce environment
2. Try the SSH plugin functions.

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)
- [ ] Optional Companion PR for updating the HappyPath tests is approved and ready to be merged

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### Happy Path Channel

HAPPY_PATH_CHANNEL=stable

### CODE REVIEW

**Comment by svor** [comment]
I suggest to wait when github-oauth plugin will be adapted to don't break the current state of the github functionality. 
@vinokurig @benoitf wdyt?

File: plugins/ssh-plugin/src/ssh-plugin.ts [hunk 1]
```diff
@@ -92,12 +89,15 @@ export class SSHPluginImpl implements SSHPlugin {
     theia.commands.registerCommand(this.cmdViewPublicKey, () => {
       this.cmdViewPublicKey.run();
     });
-    theia.commands.registerCommand(this.cmdUploadPrivateKey, () => {
-      this.cmdUploadPrivateKey.run();
-    });
-    theia.commands.registerCommand(this.cmdAddKeyToGitHub, () => {
-      this.cmdAddKeyToGitHub.run();
-    });
+    // todo remove the condition when github oAuth is available for devworkspcae engine
```
Comment (benoitf) [inline plugins/ssh-plugin/src/ssh-plugin.ts:92]:
```suggestion
    // todo remove the condition when github oAuth is available for DevWorkspace engine
```

File: plugins/ssh-plugin/src/ssh-plugin.ts [hunk 2]
```diff
@@ -109,11 +109,13 @@ export class SSHPluginImpl implements SSHPlugin {
       { label: this.cmdViewPublicKey.label },
       { label: this.cmdCreateKey.label },
       { label: this.cmdDeleteKey.label },
-      { label: this.cmdUploadPrivateKey.label },
     ];
-
-    if (gitHubActions) {
-      items.push({ label: this.cmdAddKeyToGitHub.label });
+    // todo remove the condition when github oAuth is available for devworkspcae engine
```
Comment (benoitf) [inline plugins/ssh-plugin/src/ssh-plugin.ts:113]:
```suggestion
    // todo remove the condition when github oAuth is available for DevWorkspace engine
```

File: plugins/ssh-plugin/src/ssh-plugin.ts [hunk 3]
```diff
@@ -134,10 +136,12 @@ export class SSHPluginImpl implements SSHPlugin {
       } else if (command.label === this.cmdDeleteKey.label) {
         await this.cmdDeleteKey.run({ gitCloneFlow: true });
         return true;
-      } else if (command.label === this.cmdUploadPrivateKey.label) {
+        // todo remove the DEVWORKSPACE_ID condition when github oAuth is available for devworkspcae engine
```
Comment (benoitf) [inline plugins/ssh-plugin/src/ssh-plugin.ts:139]:
```suggestion
        // todo remove the DEVWORKSPACE_ID condition when github oAuth is available for DevWorkspace engine
```

File: plugins/ssh-plugin/src/ssh-plugin.ts [hunk 4]
```diff
@@ -134,10 +136,12 @@ export class SSHPluginImpl implements SSHPlugin {
...
+      } else if (!process.env.DEVWORKSPACE_ID && command.label === this.cmdUploadPrivateKey.label) {
         await this.cmdUploadPrivateKey.run({ gitCloneFlow: true });
         return true;
-      } else if (command.label === this.cmdAddKeyToGitHub.label) {
+        // todo remove the DEVWORKSPACE_ID condition when github oAuth is available for devworkspcae engine
```
Comment (benoitf) [inline plugins/ssh-plugin/src/ssh-plugin.ts:143]: DevWorkspace
```suggestion
        // todo remove the DEVWORKSPACE_ID condition when github oAuth is available for DevWorkspace engine
```

File: plugins/ssh-plugin/src/util/ssh-secret-helper.ts [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+      const client = this.getK8sCoreApi();
+      client.defaultHeaders = { Accept: 'application/json', 'Content-Type': k8s.PatchUtils.PATCH_FORMAT_JSON_PATCH };
+      await client.patchNamespacedSecret(CREDENTIALS_SECRET_NAME, this.getWorkspaceNamespace(), patchPrivate);
+      await client.patchNamespacedSecret(CREDENTIALS_SECRET_NAME, this.getWorkspaceNamespace(), patchPublic);
+    } catch (e) {
+      theia.window.showErrorMessage('Failed to delete the SSH secret' + e);
+    }
+  }
+
+  private getWorkspaceNamespace(): string {
```
Comment (svor) [inline plugins/ssh-plugin/src/util/ssh-secret-helper.ts:120]: I think you can use `getCurrentNamespace()` method from `workspace` namespace to get Namespace

https://github.com/eclipse-che/che-theia/blob/main/extensions/eclipse-che-theia-plugin/src/che-proposed.d.ts#L32

**Comment by vinokurig** [comment]
depends on https://github.com/eclipse/che/issues/20844

**Comment by svor** [comment]
[crw-ci-test]

**Review by svor — APPROVED** [review]

File: plugins/ssh-plugin/src/util/ssh-secret-helper.ts (hunk 5 above)
Comment (vinokurig) [reply plugins/ssh-plugin/src/util/ssh-secret-helper.ts]: done

File: plugins/ssh-plugin/src/util/ssh-secret-helper.ts [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+ *
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ ***********************************************************************/
+
+import * as che from '@eclipse-che/plugin';
+// import * as fs from 'fs-extra';
```
Comment (svor) [inline plugins/ssh-plugin/src/util/ssh-secret-helper.ts:12]: probably should be removed

File: plugins/ssh-plugin/src/util/ssh-secret-helper.ts [hunk 7]
```diff
@@ -0,0 +1,171 @@
...
+        await client.patchNamespacedSecret(CREDENTIALS_SECRET_NAME, namespace, {
+          data: { ssh_config: Buffer.from(newConfigContent).toString('base64') },
+        });
+      } else {
+        await client.patchNamespacedSecret(CREDENTIALS_SECRET_NAME, namespace, {
+          data: { ssh_config: Buffer.from(configContent + keyConfig).toString('base64') },
+        });
+      }
+    } catch (e) {
+      console.error('Failed to store the the SSH system-wide config file' + e);
```
Comment (svor) [inline plugins/ssh-plugin/src/util/ssh-secret-helper.ts:162]:
```suggestion
      console.error('Failed to store the SSH system-wide config file' + e);
```

File: plugins/ssh-plugin/src/util/ssh-secret-helper.ts [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+
+  private async updateSystemConfigFile(hostName: string): Promise<void> {
+    const client = this.getK8sCoreApi();
+    const namespace = await che.workspace.getCurrentNamespace();
+    let data;
+    try {
+      const request = await client.readNamespacedSecret(CREDENTIALS_SECRET_NAME, namespace);
+      data = request.body.data;
+    } catch (e) {
+      console.error('Failed to read the the SSH secret' + e);
```
Comment (svor) [inline plugins/ssh-plugin/src/util/ssh-secret-helper.ts:140]:
```suggestion
      console.error('Failed to read the SSH secret' + e);
```

File: plugins/ssh-plugin/src/command/create-key.ts [hunk 9]
```diff
@@ -8,35 +8,37 @@
...
-    const privateKey = await theia.window.showInputBox({ placeHolder: 'Enter private key' });
+    const publicKey = (await theia.window.showInputBox({ placeHolder: 'Enter public key' })) || '';
+    const privateKey = (await theia.window.showInputBox({ placeHolder: 'Enter private key' })) || '';
 
     try {
-      await che.ssh.create({ name: hostName, service: 'vcs', publicKey: publicKey, privateKey });
-      await updateConfig(hostName);
-      await writeKey(hostName, privateKey!);
-      await theia.window.showInformationMessage(`Key pair for ${hostName} successfully created`);
+      this.sshSecretHelper.store({ name: keyName, privateKey, publicKey });
```
Comment (svor) [inline plugins/ssh-plugin/src/command/create-key.ts:38]: do we need `await` here ?

File: plugins/ssh-plugin/src/util/ssh-secret-helper.ts (hunk 6 above)
Comment (vinokurig) [reply plugins/ssh-plugin/src/util/ssh-secret-helper.ts]: fixed

File: plugins/ssh-plugin/src/command/create-key.ts (hunk 9 above)
Comment (vinokurig) [reply plugins/ssh-plugin/src/command/create-key.ts]: yes, fixed

**Review by svor — APPROVED** [review]
LGTM

**Comment by vinokurig** [comment]
depends on https://github.com/devfile/devworkspace-operator/issues/732

**Review by benoitf — APPROVED** [review]

**Pull request merged**